crates/test-utils/src/chain/nonce_manager.rs

```diff
@@ -0,0 +1,223 @@
+        let nonce = self
+            .inner
+            .get_transaction_count(self.address, block)
+            .await
+            .map_err(MiddlewareError::from_err)?;
+        self.nonce.store(nonce.as_u64(), Ordering::SeqCst);
+        self.initialized.store(true, Ordering::SeqCst);
+        Ok(nonce)
+    } // guard dropped here
+
+    /// Resets the nonce to the current transaction count. This can be used to
+    /// reset the nonce manager when the EVM snapshot is reverted.
+    pub async fn reset_nonce(&self, block: Option<BlockId>) -> Result<(), NonceManagerError<M>> {
+        // Fail if we're attempting to reset before initialization.
+        if !self.initialized.load(Ordering::SeqCst) {
+            return Err(NonceManagerError::ResetError);
+        }
+
+        // Reset the nonce to the current transaction count.
+        let nonce = self
+            .inner
+            .get_transaction_count(self.address, block)
+            .await
+            .map_err(MiddlewareError::from_err)?;
+        self.nonce.store(nonce.as_u64(), Ordering::SeqCst);
+
+        Ok(())
+    }
+
+    async fn get_transaction_count_with_manager(
+        &self,
+        block: Option<BlockId>,
+    ) -> Result<U256, NonceManagerError<M>> {
+        // initialize the nonce the first time the manager is called
+        if !self.initialized.load(Ordering::SeqCst) {
+            let nonce = self
+                .inner
+                .get_transaction_count(self.address, block)
+                .await
+                .map_err(MiddlewareError::from_err)?;
+            self.nonce.store(nonce.as_u64(), Ordering::SeqCst);
+            self.initialized.store(true, Ordering::SeqCst);
+        }
+
+        Ok(self.next())
+    }
+}
+
+#[derive(Error, Debug)]
+/// Thrown when an error happens at the Nonce Manager
+pub enum NonceManagerError<M: Middleware> {
+    /// Thrown when the nonce manager fails to be reset. This occurs when reset
+    /// is called before initialization.
+    #[error("ResetError")]
+    ResetError,
+    /// Thrown when the nonce manager exceeds the maximum number of retries
+    /// while submitting transactions.
+    #[error("RetryError")]
+    RetryError,
+    /// Thrown when the internal middleware errors
+    #[error("{0}")]
+    MiddlewareError(M::Error),
+}
+
+impl<M: Middleware> MiddlewareError for NonceManagerError<M> {
+    type Inner = M::Error;
+
+    fn from_err(src: M::Error) -> Self {
+        NonceManagerError::MiddlewareError(src)
+    }
+
+    fn as_inner(&self) -> Option<&Self::Inner> {
+        match self {
+            NonceManagerError::ResetError => None,
+            NonceManagerError::RetryError => None,
+            NonceManagerError::MiddlewareError(e) => Some(e),
+        }
+    }
+}
+
+#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
+#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
+impl<M> Middleware for NonceManagerMiddleware<M>
+where
+    M: Middleware,
```

Comment:
I was trying to make sense of how the Middleware pattern works in Rust. So `Middleware` is a trait and this implements it on the generic struct which takes an `inner` that also implements the `Middleware` trait?

Reply:
Yeah, that's the idea. It's a rusty version of the old Metamask provider engine. The basic idea is that every layer of the stack can alter the execution of a particular piece of the Ethereum JSON RPC. For example, I could make a nonce manager which handles nonces in a sane way across concurrent uses, I could write a cacher that caches calls to `getPoolInfo`, etc. The way that Rust does this is by including the `Middleware` trait which ensures that requests can get bubbled down the provider stack.